xbmc/cores/RetroPlayer/cheevos/Cheevos.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ *  Copyright (C) 2012-2020 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "Cheevos.h"
+#include "filesystem/File.h"
+#include "games/addons/GameClient.h"
+#include "URL.h"
+#include "utils/auto_buffer.h"
+#include "utils/JSONVariantParser.h"
+#include "utils/log.h"
+#include "utils/URIUtils.h"
+#include "utils/Variant.h"
+
+using namespace KODI;
+using namespace RETRO;
+
+// API JSON Field names
+constexpr auto SUCCESS = "Success";
+constexpr auto GAME_ID = "GameID";
+constexpr auto PATCH_DATA = "PatchData";
+constexpr auto RICH_PRESENCE = "RichPresencePatch";
+
+constexpr int URL_SIZE = 512;
+
+CCheevos::CCheevos(GAME::CGameClient* gameClient, const std::string userName, const std::string loginToken)
+  : m_gameClient(gameClient),
+    m_userName(userName),
+    m_loginToken(loginToken)
+{
+}
+
+bool CCheevos::LoadData()
+{
+  const std::string extension = URIUtils::GetExtension(m_gameClient->GetGamePath());
+  auto it = m_extensionToConsole.find(extension);
+
+  if (it == m_extensionToConsole.end())
+    return false;
+
+  if (m_romHash.empty())
+  {
+
+    char hash[33];
+    if (!m_gameClient->RCGenerateHashFromFile(hash, it->second,
+                                              m_gameClient->GetGamePath().c_str()))
+    {
+      return false;
+    }
+
+    m_romHash = hash;
+  }
+
+  char requestURL[URL_SIZE];
+
+  if (!m_gameClient->RCGetGameIDUrl(requestURL, URL_SIZE, m_romHash.c_str()))
+    return false;
+
+  XFILE::CFile response;
+  response.CURLCreate(requestURL);
+  response.CURLOpen(0);
+
+  char responseStr[50];
```

Comment:
magic number? where does 50 come from? should it be declared next to `URL_SIZE`?